### PR TITLE
Start moving genericapiserver to staging

### PR DESCRIPTION
**Pull request by deads2k**
This moves `pkg/auth/user` to `staging/k8s.io/genericapiserver/pkg/authentication/user`.  I'll open a separate pull into the upstream gengo to support using `import-boss` on vendored folders to support staging.

After we agree this is the correct approach and see everything build, I'll start moving other packages over which don't have k8s.io/kubernetes deps.

@kubernetes/sig-api-machinery-misc @lavalamp 
@sttts @caesarxuchao ptal

### CODE REVIEW

**Comment by deads2k** [comment]
@krousey @mikedanese one of you was the bazel guy, right?  Can you help me figure out what has gone wrong here: https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/pr-logs/pull/39384/pull-kubernetes-bazel/6287297323453159316/ .  `hack/update-bazel.sh` is happy on my local machine.

**Comment by deads2k** [comment]
opened https://github.com/kubernetes/kubernetes/pull/39387 to split out the import-boss updates from this pull.

File: staging/src/k8s.io/genericapiserver/pkg/authentication/user/doc.go [hunk 1]
```diff
@@ -16,4 +16,4 @@ limitations under the License.
 
 // Package user contains utilities for dealing with simple user exchange in the auth
 // packages. The user.Info interface defines an interface for exchanging that info.
-package user // import "k8s.io/kubernetes/pkg/auth/user"
+package user // import "k8s.io/genericapiserver/pkg/authentication/user"
```
Comment (deads2k) [inline staging/src/k8s.io/genericapiserver/pkg/authentication/user/doc.go]: @lavalamp The pull looks massive, but this is the meat right here.  Moving three files.

Comment (lavalamp) [inline staging/src/k8s.io/genericapiserver/pkg/authentication/user/doc.go]: Gotcha. Is `k8s.io/genericapiserver` set in stone? Just thinking about hypothetical users seeing this in the repo list. `k8s.io/apiserver-lib`? Or just plain `k8s.io/apiserver` with the understanding that the specific binaries will all be prefixed with something (e.g., `kube-apiserver`)

File: staging/src/k8s.io/genericapiserver/.import-restrictions [hunk 2]
```diff
@@ -0,0 +1,10 @@
+{
+  "Rules": [
+    {
+      "SelectorRegexp": "k8s[.]io/kubernetes",
+      "ForbiddenPrefixes": [
```
Comment (lavalamp) [inline staging/src/k8s.io/genericapiserver/.import-restrictions]: Forbid everything-- awesome, thanks.

**Comment by lavalamp** [comment]
I am LGTM on this, I just want to make sure we've thought about the name carefully.

File: staging/src/k8s.io/genericapiserver/pkg/authentication/user/doc.go (hunk 1 above)
Comment (deads2k) [inline staging/src/k8s.io/genericapiserver/pkg/authentication/user/doc.go]: I'm fine with k8s.io/apiserver.  I think we have until 1.6 dcut to change our mind, so I'll update tomorrow.

Comment (sttts) [inline staging/src/k8s.io/genericapiserver/pkg/authentication/user/doc.go]: Prefering `pkg/apiserver` as well.

File: hack/godep-save.sh [hunk 3]
```diff
@@ -40,6 +40,7 @@ pushd "${KUBE_ROOT}" > /dev/null
   # let other packages use the staging client as if it were vendored.
   if [ ! -e "vendor/k8s.io/client-go" ]; then
     ln -s ../../staging/src/k8s.io/client-go vendor/k8s.io/client-go
+    ln -s ../../staging/src/k8s.io/genericapiserver vendor/k8s.io/genericapiserver
```
Comment (sttts) [inline hack/godep-save.sh]: put this into its own if block.

**Comment by deads2k** [comment]
@sttts I've updated for the new name and rebased.  I also split it into meat and mechanicals.

File: staging/src/k8s.io/apiserver/pkg/authentication/user/doc.go [hunk 4]
```diff
@@ -16,4 +16,4 @@ limitations under the License.
 
 // Package user contains utilities for dealing with simple user exchange in the auth
 // packages. The user.Info interface defines an interface for exchanging that info.
-package user // import "k8s.io/kubernetes/pkg/auth/user"
```
Comment (sttts) [inline staging/src/k8s.io/apiserver/pkg/authentication/user/doc.go]: Do we want to change this style?

**Comment by sttts** [comment]
Cannot judge bazel changes. Otherwise, lgtm.

File: staging/src/k8s.io/genericapiserver/pkg/authentication/user/doc.go (hunk 1 above)
Comment (smarterclayton) [inline staging/src/k8s.io/genericapiserver/pkg/authentication/user/doc.go]: Prefer k8s.io/apiserver

**Comment by deads2k** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by deads2k** [comment]
@k8s-bot gce etcd3 e2e test this

**Comment by fejta** [comment]
@k8s-bot test this

kubernetes/test-infra#1486

**Comment by deads2k** [comment]
@k8s-bot unit test this: issue #38518

**Pull request merged**